plugin/impl.go

```diff
@@ -0,0 +1,225 @@
+		return fmt.Errorf("repository named %s must use %s as its source", nugetOrgName, nugetOrgSource)
+	}
+	if p.settings.Name != nugetOrgName && p.settings.Source == nugetOrgSource {
+		return fmt.Errorf("source %s must use %s as its repository name", nugetOrgSource, nugetOrgName)
+	}
+	if _, err := url.Parse(p.settings.Source); err != nil {
+		return fmt.Errorf("could not parse source url %s: %w", p.settings.Source, err)
+	}
+
+	logrus.WithFields(logrus.Fields{
+		"name":   p.settings.Name,
+		"source": p.settings.Source,
+	}).Info("Using NuGet repository")
+
+	return nil
+}
+
+// Execute provides the implementation of the plugin.
+func (p *Plugin) Execute() error {
+	cmds := []*exec.Cmd{
+		versionCmd(),
+	}
+
+	if p.settings.Name != nugetOrgName {
+		cmds = append(cmds, addSourceCmd(p.settings.Source, p.settings.Name))
+	}
+
+	cmds = append(cmds, listSourcesCmd())
+	cmds = append(cmds, pushPackageCmd(p.settings.nupkg, p.settings.Name, p.settings.APIKey))
+
+	return runCommands(cmds, "")
+}
+
+// nupkgFromNuspec reads the nuspec file and determines the filename.
+func nupkgFromNuspec(file string) (string, error) {
+	if !fileExists(file) {
+		return "", fmt.Errorf(".nuspec file not found at %s", file)
+	}
+
+	// Read the file
+	nuspec, err := os.Open(file)
+	if err != nil {
+		return "", fmt.Errorf("unable to open .nuspec file %s: %w", file, err)
+	}
+
+	bytes, err := ioutil.ReadAll(nuspec)
+	if err != nil {
+		return "", fmt.Errorf("unable to read .nuspec file %s: %w", file, err)
+	}
+
+	// Unmarshal the file
+	var doc struct {
+		XMLName  xml.Name       `xml:"package"`
+		Metadata nuspecMetadata `xml:"metadata"`
+	}
+	if err := xml.Unmarshal(bytes, &doc); err != nil {
+		return "", fmt.Errorf("unable to parse .nuspec file %s: %w", file, err)
+	}
+
+	logrus.WithFields(logrus.Fields{
+		"name":    doc.Metadata.Name,
+		"version": doc.Metadata.Version,
+	}).Info("Found .nupsec file")
+
+	nupkgName := fmt.Sprintf("%s.%s.nupkg", doc.Metadata.Name, doc.Metadata.Version)
+
+	return path.Join(path.Dir(file), nupkgName), nil
+}
+
+// fileExists determines if the file is present.
+func fileExists(file string) bool {
+	info, err := os.Stat(file)
+	if os.IsNotExist(err) {
+		return false
+	}
+
+	return !info.IsDir()
+}
+
+// versionCmd gets the nuget version.
+func versionCmd() *exec.Cmd {
+	return exec.Command("dotnet", "nuget", "--version")
+}
+
+// listSourcesCmd lists the nuget repositories.
+func listSourcesCmd() *exec.Cmd {
+	return exec.Command("dotnet", "nuget", "list", "source")
+}
+
+// addSourceCmd creates a new nuget repository source.
+func addSourceCmd(source, name string) *exec.Cmd {
+	return exec.Command("dotnet", "nuget", "add", "source", source, "--name", name)
+}
+
+// pushPackageCmd pushes a package to the nuget repository.
+func pushPackageCmd(path, name, key string) *exec.Cmd {
+	return exec.Command("dotnet", "nuget", "push", path, "--source", name, "--api-key", key)
```

Comment:
Trace will print the api key?

Reply:
Its my understanding that if its a secret it'll end up being hidden in the log.

Reply:
With npm I was able to just create a config file that had the credentials. With Nuget the api key is always encrypted if stored in a file so its not possible for me to just replicate that.

I don't know if we would want something like a `cmd` library in the plugin lib. Most of this stuff was copied from `drone-npm`. If that's the case then we can add something that'll handle the case where we would want to control what is printed out directly.